design/pool-configuration.md

```diff
@@ -0,0 +1,533 @@
+
+The `BGPAdvertisement` entity represents the intention to advertise a given IP via the BGP Protocol. It contains properties related to the type of advertisement and a reference to the address pools we want we advertise. This initial version matches the current behaviour, where an IP is advertised to all the `BGPPeer`s configured in MetalLB.
+
+```go
+type bgpAdvertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+Open question: I would like to preserve not having to specify all the ips / peers in order to match the current "ease of configuration" of metallb. One possible way is to assume that all the pools are advertised to all the peers if no selection is specified, the other one is to provide an example using the label selector.
+
+#### L2 Advertisement
+
+The `L2Advertisement` entity represents the intention to advertise a given IP via L2. Currently, there are no knobs related to L2 advertisement.
+
+```go
+type l2Advertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+```
+
+### BGP Peer
+
+The `BgpPeer` entity represents an external BGP peer we want to establish a session with. It contains the properties of the sessions, and it's a 1:1 mapping of the current bgp peer structure without the node selector. We also add the `bfdProfile` structure here, as it is part of the session definition.
+
+```go
+type bgpPeer struct {
+  Name              string
+  MyASN         uint32
+  ASN           uint32
+  Addr          string
+  SrcAddr       string
+  Port          uint16
+  HoldTime      string
+  KeepaliveTime string
+  RouterID      string
+  Password      string
+  BFDProfile    string
+  EBGPMultiHop  bool
+}
+
+type bfdProfile struct {
+  Name             string  
+  ReceiveInterval  *uint32 
+  TransmitInterval *uint32 
+  DetectMultiplier *uint32 
+  EchoInterval     *uint32 
+  EchoMode         bool    
+  PassiveMode      bool    
+  MinimumTTL       *uint32 
+}
+```
+
+Example CRDs:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPPeer
+metadata:
+  name: peer-bfd
+  namespace: metallb-system
+spec:
+  peerAddress: 10.0.0.2
+  peerASN: 64501
+  myASN: 64500
+  routerID: 10.10.10.10
+  peerPort: 1
+  holdTime: "10s"
+  sourceAddress: 1.1.1.1
+  bfdProfile: bfdprofilefull
```

Comment:
What's the reason for a separate `BFDProfile` resource?  Is it possible to associate a `BFDProfile` with multiple BGP peers?

Reply:
Yes, or to put it the other way around, multiple bgp peers can use the same profile. It's also the same way frr handles it, which is not a good enough reason BUT with a 1:1 mapping between a peer and its bfd settings, we'd have to create multiple bfd profiles in the frr configuration (and this does seemed reasonable).

Reply:
Gotcha!  Yeah, I'm just trying to figure out if there's a way to prevent config sprawl.